discover/transport.go

```diff
@@ -0,0 +1,318 @@
+	RPL_MOTD = "378"
+)
+
+// Transport interface is used to create an abstraction layer over
+// system-dependable representation of discovery service messages (IRC). It
+// gives the ability to discovery service operate only with final representation
+// of messages like ChanAnnouncement, NodeAnnouncement thereby decouple
+// this systems and give them ability to be easily testable.
+type Transport interface {
+	Connect() (string, error)
+
+	// Joined returns the channel which will be triggered when some new
+	// peer joined.
+	Joined() chan bool
+
+	// Send is used to send the discovery message.
+	Send(interface{}) error
+
+	// Receive return the channel which will be triggered when new
+	// discovery message is received.
+	Receive() chan interface{}
+
+	// Decode is used to convert transport message into discovery
+	// service messages.
+	Decode(string) (interface{}, error)
+
+	// Decode is used to convert discovery service messages into
+	// transport messages.
+	Encode(interface{}) (string, error)
+}
+
+var (
+	defaultServer           = "localhost:6667"
+	defaultTimeout          = time.Second * 15
+	defaultLightningChannel = "#lightning-nodes"
+	defaultDataSeparator    = " "
+	defaultMaxAliasLen      = 64
+)
+
+// IRCTransport is an implementation of discovery message transport which
+// uses IRC bot and IRC channels as carrier of announcement data.
+type IRCTransport struct {
+	quit chan bool
+
+	hostname chan string
+	incoming chan interface{}
+	joined   chan bool
+
+	conn *client.Conn
+
+	// Public and private keys of peer node, used to sign data.
+	pubKey  *btcec.PublicKey
+	privKey *btcec.PrivateKey
+}
+
+var _ Transport = (*IRCTransport)(nil)
+
+func NewIRCTransport(privKey *btcec.PrivateKey) *IRCTransport {
+	t := &IRCTransport{
+		quit:     make(chan bool),
+		incoming: make(chan interface{}),
+		joined:   make(chan bool),
+		hostname: make(chan string),
+		privKey:  privKey,
+		pubKey:   privKey.PubKey(),
+	}
+
+	cfg := client.NewConfig(t.getNick())
+	cfg.Timeout = defaultTimeout
+	conn := client.Client(cfg)
+
+	// Initialize IRC message handlers.
+	conn.HandleFunc(client.CONNECTED, t.onConnected)
+	conn.HandleFunc(client.DISCONNECTED, t.onDisconnected)
+	conn.HandleFunc(client.PRIVMSG, t.onPrivMsg)
+	conn.HandleFunc(client.JOIN, t.onJoin)
+	conn.HandleFunc(RPL_MOTD, t.onWhoisResponse)
+
+	t.conn = conn
+	return t
+}
+
+// getNick generates the IRC channels nickname.
+func (t *IRCTransport) getNick() string {
+	hexPubKey := t.pubKey.SerializeCompressed()
+	return fmt.Sprintf("N%.12s", hex.EncodeToString(hexPubKey))
+}
+
+func (t *IRCTransport) Connect() (string, error) {
+	if err := t.conn.ConnectTo(defaultServer); err != nil {
+		return "", fmt.Errorf("Connection error: %s\n", err.Error())
+	}
+
+	if ip, ok := <-t.hostname; !ok {
+		return "", errors.New("Can't receive ip address.")
+	} else {
+		return ip, nil
+	}
+}
+
+func (t *IRCTransport) Decode(data string) (interface{}, error) {
```

Comment:
In the same vein as the comment above, I don't think this needs to be a publicly exported part of the interface. Instead (since this is only called from internal functions), if can just be a transport implementation specific internal encoding/decoding function.